pkg/internalregistry/controllers/keyid_observation_controller.go

```diff
@@ -0,0 +1,145 @@
+package controllers
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	informers "k8s.io/client-go/informers/core/v1"
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/keyutil"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type keyIDObservation struct {
+	secrets       listers.SecretLister
+	cacheSyncs    []cache.InformerSynced
+	queue         workqueue.RateLimitingInterface
+	observeKeyIDs func(keyIDs []string) error
+}
+
+func NewKeyIDObservationController(secrets informers.SecretInformer, keyIDsObservationFunc func(keyIDs []string) error) *keyIDObservation {
+	c := &keyIDObservation{
+		secrets:       secrets.Lister(),
+		observeKeyIDs: keyIDsObservationFunc,
+		cacheSyncs:    []cache.InformerSynced{secrets.Informer().HasSynced},
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "sa-signing-key-secrets"),
+	}
+	secrets.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj any) bool {
+			if secret, ok := obj.(*corev1.Secret); ok {
+				return (secret.Namespace == "openshift-kube-apiserver") && (secret.Name == "bound-service-account-signing-key")
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj any) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					c.queue.Add(key)
+				}
+			},
+			UpdateFunc: func(_, obj any) {
+				key, err := cache.MetaNamespaceKeyFunc(obj)
+				if err == nil {
+					c.queue.Add(key)
+				}
+			},
+		},
+	})
+	return c
+}
+
+func (c *keyIDObservation) sync(ctx context.Context, key string) error {
+	klog.V(4).InfoS("sync", "key", key)
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	secret, err := c.secrets.Secrets(ns).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	pem, ok := secret.Data["service-account.pub"]
+	if !ok {
+		return fmt.Errorf("expected data service-account.pub not found")
+	}
+	keys, err := keyutil.ParsePublicKeysPEM(pem)
+	if err != nil {
+		return err
+	}
+	// compute key ID (e.g. hash) for token signing private keys from the public keys
+	var kids []string
+	for _, key := range keys {
+		der, err := x509.MarshalPKIXPublicKey(key)
+		if err != nil {
+			return err
+		}
+		hashFunc := crypto.SHA256.New()
+		hashFunc.Reset()
+		_, err = hashFunc.Write(der)
+		if err != nil {
+			return err
+		}
+		kids = append(kids, base64.RawURLEncoding.EncodeToString(hashFunc.Sum(nil)))
+	}
+	if len(kids) == 0 {
+		// not expecting this, don't clobber existing values
+		klog.V(2).InfoS("no key IDs observed")
+		return nil
+	}
```

Comment:
This would leave the controller in a state where it could not generate a useful pull secret. I moved this logic into the imagepullsecretcontroller where I think it makes more sense.